Devscord.DiscordFramework/Framework/Commands/Services/BotCommandsMatchingService.cs

```diff
@@ -1,87 +1,146 @@
 
 namespace Devscord.DiscordFramework.Framework.Commands.Services
 {
     public class BotCommandsMatchingService
     {
+        private readonly Regex _exTime = new Regex(@"\d+(ms|d|h|m|s)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+        private readonly Regex _exUserMention = new Regex(@"\<@!?\d+\>", RegexOptions.Compiled);
+        private readonly Regex _exChannelMention = new Regex(@"\<#\d+\>", RegexOptions.Compiled);
 
-        private readonly Regex _exTime = new Regex(@"\d+(h|m|s)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
-        private readonly Regex _exUserMention = new Regex(@"\<@[^\d]*\d+\>", RegexOptions.Compiled);
-        private readonly Regex _exChannelMention = new Regex(@"\<#[^\d]*\d+\>", RegexOptions.Compiled);
-
-        public bool IsMatchedWithCommand(DiscordRequest request, BotCommandTemplate template)
+        public bool IsDefaultCommand(BotCommandTemplate template, IEnumerable<DiscordRequestArgument> arguments, bool isCommandMatchedWithCustom)
         {
-            if (!request.IsCommandForBot)
+            var requiredProperties = template.Properties.Where(x => !x.IsOptional);
+            if (requiredProperties.Any(property => arguments.All(arg => property.Name.ToLowerInvariant() != arg.Name?.ToLowerInvariant())))
             {
                 return false;
             }
-            if (request.Name.ToLowerInvariant() != template.NormalizedCommandName)
+            if (!isCommandMatchedWithCustom)
             {
-                return false;
+                return true;
             }
-            if (!this.CompareArgumentsToProperties(request.Arguments.ToList(), template.Properties.ToList()))
+            // if we give the command in such a way that isCommandMatchedWithCustom is true and it has been checked before that all the required arguments have been given and it turned out that they were given in this way, 
+            // the program checks whether all given arguments (even the optional ones) are known to the command, so it is known whether we are dealing with a command in the default form or in the form of custom.
+            // if the method returns false after the following operations, it means that the command is in a custom form
+            var argsWithoutListSubarguments = arguments
+                .Select(x => x.Name?.ToLowerInvariant())
+                .ToList();
+            var lists = template.Properties
+                .Where(property => property.Type == BotCommandPropertyType.List)
+                .Select(x => x.Name.ToLowerInvariant())
+                .ToList();
+            for (int i = 0; i < argsWithoutListSubarguments.Count; i++)
             {
-                return false;
+                if (!lists.Contains(argsWithoutListSubarguments[i]))
+                {
+                    continue;
+                }
+                while (i + 1 < argsWithoutListSubarguments.Count && argsWithoutListSubarguments[i + 1] == null)
```

Comment:
`i + 1 < argsWithoutListSubarguments.Count`
szczegół, ale myśle że czytelniej `i < argsWithoutListSubarguments.Count - 1`

Reply:
to zależy jak kto patrzy, bo ja widzę to w ten sposób, sprawdzam czy następny element nie jest poza przedziałem, a w twoim jest sprawdzane czy obecny element jest w tym przedziale. Ja bym to zostawił.